### ISSUE TITLE
gradients of index_select wrong on GPU

### ISSUE THREAD

**Issue by mys007** [pytorch/pytorch#3644]
I've found out that gradient of `index_select` (i..e `index_add`) behaves non-deterministically - and wrong - on GPU for some rather random conditions. The following code illustrates the problem:

```python
import numpy as np
import torch
from torch.autograd import Variable

def set_seed(seed):
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)

def main():
    set_seed(1)
    cuda = True
    n, e, c = 5, 100, 5
    input = torch.randn(n,c)
    idxn = torch.from_numpy(np.random.randint(n,size=e))  # indices are repeated

    if cuda:
        input = input.cuda(); idxn = idxn.cuda()

    gradsI, gradsS = [], []
    N = 4
    for i in range(N):
        inputv = Variable(input, requires_grad=True)
        sel_input = torch.index_select(inputv, 0, Variable(idxn))
        sel_input.retain_grad()

        # the following computation is one of the "random conditions"
        data = [torch.sum(sel_input.narrow(0,0,e//2), 0),
                torch.sum(sel_input.narrow(0,e//2,e//2), 0) ]
        out = torch.cat(data,0)
        out.exp().sum().backward()

        gradsI.append(inputv.grad.data.cpu().clone())
        gradsS.append(sel_input.grad.data.cpu().clone())

    for i in range(N):
        for j in range(N):
            print(i,j,(gradsI[i]-gradsI[j]).abs().max(), (gradsS[i]-gradsS[j]).abs().max())
```

If `cuda=False`, everything works and 0s are printed, i.e. all gradients are equal. However, with cuda on many nonzero differences in `gradsI` but not in `gradsS` pop up, sometimes of high magnitude, e.g.:
```
0 0 0.0 0.0
0 1 1.4411518807585587e+17 0.0
0 2 1.4411518807585587e+17 0.0
0 3 1.4411518807585587e+17 0.0
1 0 1.4411518807585587e+17 0.0
1 1 0.0 0.0
1 2 3.602879701896397e+16 0.0
1 3 0.015625 0.0
2 0 1.4411518807585587e+17 0.0
2 1 3.602879701896397e+16 0.0
2 2 0.0 0.0
2 3 3.602879701896397e+16 0.0
3 0 1.4411518807585587e+17 0.0
3 1 0.015625 0.0
3 2 3.602879701896397e+16 0.0
3 3 0.0 0.0
````

This indicates that there is something wrong with backward of `torch.index_select`. Setting `torch.set_default_tensor_type('torch.DoubleTensor')` doesn't help. Can anybody shed light onto this?

**Comment by zou3519** [comment]
@mys007 I believe this is a rounding error.

I took your code and modified it slightly to print out one entry (0, 2) of the gradsI[i]-gradsI[j] matrix that seems to be responsible for the large magnitude.
https://gist.github.com/zou3519/de11aa86df07712784c5009ce0362787

The output is:
```
-------------------
0 0 0.0 0.0
9.737174387686579e+23 9.737174387686579e+23
-------------------
0 1 1.4411518807585587e+17 0.0
9.737174387686579e+23 9.737175108262519e+23
-------------------
1 0 1.4411518807585587e+17 0.0
9.737175108262519e+23 9.737174387686579e+23
-------------------
1 1 0.0 0.0
9.737175108262519e+23 9.737175108262519e+23
```
(the two numbers in the second row of each group are `(gradsI[i]-gradsI[j])[0, 2]`).
Because we're doing floating point math, we can expect the numbers to be accurate to 7 digits, which they are. However, subtracting the numbers gives a difference of something that is ~1e17, as you noticed.

Changing your code to use double tensors (I changed the line `input = torch.randn(n, c)` to `input = torch.randn(n, c).double()) gives better results for me:
```
-------------------
0 0 0.0 0.0
9.737195888283742e+23 9.737195888283742e+23
-------------------
0 1 134217728.0 0.0
9.737195888283742e+23 9.737195888283741e+23
-------------------
1 0 134217728.0 0.0
9.737195888283741e+23 9.737195888283742e+23
-------------------
1 1 0.0 0.0
9.737195888283741e+23 9.737195888283741e+23
```
There's still a difference but the double math has made it more precise.

**Comment by mys007** [comment]
I agree with your analysis. However, 1) float arithmetics should work the same on gpu and cpu, and 2) values like 9.737195888283742e+23 shouldn't be produced by backprop there at the first place.

**Comment by soumith** [comment]
Yea the thing to investigate is why there are such large values in gradients when given overlapping indices.

**Comment by zou3519** [comment]
Hmm okay, that makes sense!

Regarding the large gradients: I numerically computed the gradient by perturbing input[0, 2] by a small amount and looking at how much it changes the output and it looks like something around 1e+24 is reasonable for the gradient. The output seems to be high because of the .exp() operation (one of the inputs to that was 41 from a quick glance). I'll look more into this tomorrow

**Comment by mys007** [comment]
@zou3519 I'm not exactly sure how did you get the values. In the original code, `input` is the same for all iterations and the values in `input` are normal with no large outliers.

**Comment by mys007** [comment]
@zou3519 Indeed, I used exp for this blowing-up purpose, which is however fairly common if you use softmax and train to convergence. But the effect also happens if you'll have `input = torch.randn(n,c) * 0.01`, albeit one doesn't reach such high exponents.

not a problem of the range

**Comment by zou3519** [comment]
So let `output = out.exp().sum()`. I'm going to numerically compute the partial derivative of output with respect to `input[0, 2]`, which is theoretically equal to `gradsI[i][0, 2]` for all `i`.

Here's the code I used to do that: https://gist.github.com/zou3519/4cb1c3a674e9785bf8121e851d4ded0d
I ran this twice, once with the line `input[0, 2] += 0`, and once with the `input[0, 2] += 0.1`.
They gave the following output

```
output =
 9.7372e+22
[torch.cuda.FloatTensor of size 1 (GPU 0)]
  input =  -0.43049895763397217
output =
 2.6468e+23
[torch.cuda.FloatTensor of size 1 (GPU 0)]
  input =  -0.33049896359443665
```
This gives two data points (input, output): `(-0.43049895763397217, 9.7372e+22)` and `(-0.33049896359443665, 2.6468e+23)`. Estimating the partial derivative with (change in output) / (change in input) gives 1.67308e+24, so I think it's reasonable that given the forward pass, the gradients are very large.

**Comment by soumith** [comment]
> which is however fairly common if you use softmax and train to convergence. 

This is not common at all. For precisely numerical stability issues, one uses log + sum + exp formulations like log_softmax + NLLLoss or bce_loss_with_logits, or calling [log_sum_exp](https://github.com/pytorch/pytorch/issues/2591#issuecomment-338980717).

This numerical issue is less about index_select's gradient and more about the intermediate ops being numerically unstable for the gradient.

@mys007 let me know if you disagree.

**Comment by mys007** [comment]
I indeed do disagree, because the core issue remains unanswered. @zou3519 showed that the gradient of exp() function is correct and large, given the input, fair enough. But I chose this example on purpose to demonstrate that the scale of non-determinism in `index_add` depends on the scale of gradients. My issue is valid even if I write `input = torch.randn(n,c) * 0.01` - the gradients are then small but **still different between runs**.

**Comment by soumith** [comment]
non-determinism is understandable, index_add uses atomicAdd. Writing an index_add kernel that is deterministic involves a two-pass algorithm (and a temporary buffer), which sucks for perf.

**Comment by mys007** [comment]
Well, I see. Nevertheless, it would be super handy to have a list of non-deterministic functions somewhere.

**Comment by andrei-pokrovsky** [comment]
Would it perhaps make sense to add an atomic_add=False option to index_select?
